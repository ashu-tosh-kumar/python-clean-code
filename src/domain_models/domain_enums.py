from enum import Enum


class ResponseStatus(Enum):
    """API response status"""

    SUCCESS = "SUCCESS"
    ERROR = "ERROR"
