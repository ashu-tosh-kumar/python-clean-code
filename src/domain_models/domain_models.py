from pydantic import BaseModel


class StackOverflowQuestion(BaseModel):
    """Represents a Stack Overflow question"""

    no: int
    text: str
    comments: str | None = None


class StackOverflowAnswer(BaseModel):
    """Represents a Stack Overflow answer"""

    text: str
    accepted: bool
