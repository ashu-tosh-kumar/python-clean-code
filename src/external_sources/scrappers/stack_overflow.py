import requests
from bs4 import BeautifulSoup

from src.constants import mock_data
from src.db_wrappers.in_memory_db import db
from src.domain_models import domain_models
from src.external_sources.external_source_base import ExternalSource
from src.initializer import logger


class StackOverflowException(Exception):
    """Raised if any exception in `StackOverflow`"""

    pass


class StackOverflow(ExternalSource):
    """Scrapper for StackOverflow"""

    def __init__(self) -> None:
        self._target_url = "https://stackoverflow.com/"

        logger.debug("Initialized the StackOverflow Scrapper")

    def _fetch_page(self, target_url: str) -> bytes:
        """Fetches the page of the given url `target_url` using GET HTTP call

        Args:
            target_url (str): Target url that needs to be fetched

        Returns:
            bytes: Returns the page content of the `target_url`
        """
        logger.info(f"fetching page {target_url}")
        page = requests.get(target_url)

        if page.status_code != 200:
            error_msg: str = f"Received non 200 status code in fetching {target_url}"
            logger.exception(error_msg)
            raise StackOverflowException(error_msg)

        return page.content

    def _make_soup(self, bytes_content: bytes) -> BeautifulSoup:
        """Creates a Beautiful soup for the given page content

        Args:
            bytes_content (bytes): Page content for which soup needs to be created

        Returns:
            BeautifulSoup: Returns Beautiful soup for given page content `bytes_content`
        """
        logger.info("Making soup out of bytes content")
        logger.debug(f"Making soup out of bytes content: {bytes_content}")
        soup = BeautifulSoup(bytes_content, "html.parser")

        return soup

    def _parse_page(self, soup: BeautifulSoup) -> list[tuple[domain_models.StackOverflowQuestion, domain_models.StackOverflowAnswer]]:
        """Parses the page using the given soup `soup` and returns parsed questions and
        answers

        Args:
            soup (BeautifulSoup): Beautiful soup that needs to be parsed

        Returns:
            list[tuple[domain_models.StackOverflowQuestion,
            domain_models.StackOverflowAnswer]]: Returns a list of tuples where each
            tuple represents one question and one answer
        """
        logger.info("Parsing data out of the soup")
        logger.debug(f"Parsing data out of the soup: {soup}")
        questions_answers: list[tuple[domain_models.StackOverflowQuestion, domain_models.StackOverflowAnswer]] = []

        # NOTE: Core logic to parse our hypothetical StackOverflow website would come
        # here.
        logger.info(f"Parsing the soup: {soup}")

        # For the sake of understanding and example, we will use mocked data here.
        logger.info(f"Using mocked data: {mock_data.MOCK_STACK_OVERFLOW_DATA}")
        for mock_ques, mock_ans in mock_data.MOCK_STACK_OVERFLOW_DATA:
            stack_question = domain_models.StackOverflowQuestion(**mock_ques)
            stack_answer = domain_models.StackOverflowAnswer(**mock_ans)
            questions_answers.append((stack_question, stack_answer))

        return questions_answers

    def _store_into_db(self, questions_answers: list[tuple[domain_models.StackOverflowQuestion, domain_models.StackOverflowAnswer]]) -> None:
        """Stores the given list of question and answers into database

        Args:
            questions_answers (list[tuple[domain_models.StackOverflowQuestion,
            domain_models.StackOverflowAnswer]]): list of tuples where each tuple
            represents one question and one answer
        """
        logger.info("Storing the list of questions and answers into database")
        logger.debug(f"Storing the list of questions and answers into database: {questions_answers}")
        for question, answer in questions_answers:
            db.upsert_question_answer(
                ques_no=question.no, ques_text=question.text, ques_comments=question.comments, ans_text=answer.text, ans_accepted=answer.accepted
            )

    def run(self) -> None:
        """Main run method responsible for running the scrapers and storing the scraped
        data into the database
        """
        page = self._fetch_page(self._target_url)
        soup = self._make_soup(page)
        questions_answers = self._parse_page(soup)
        self._store_into_db(questions_answers)
