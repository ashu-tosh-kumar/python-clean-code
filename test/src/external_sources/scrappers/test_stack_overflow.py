from test.src.external_sources.scrappers.stack_overflow_page_bytes_content import (
    stack_overflow_page_bytes_content,
)
from unittest.mock import Mock

import pytest
from bs4 import BeautifulSoup

from src.domain_models.domain_models import StackOverflowAnswer, StackOverflowQuestion
from src.external_sources.scrappers.stack_overflow import (
    StackOverflow,
    StackOverflowException,
)

DUMMY_TARGET_URL = "https://unittest-url"
REQUESTS = "src.external_sources.scrappers.stack_overflow.requests"


@pytest.fixture
def stack_overflow():
    yield StackOverflow()


class TestStackOverflowException:
    def test_stack_overflow_exception_should_exist(self):
        with pytest.raises(StackOverflowException):
            raise StackOverflowException()


class TestStackOverflow:
    def test__fetch_page_should_fetch_target_url(self, mocker, stack_overflow: StackOverflow):
        stub_requests = mocker.patch(REQUESTS)
        dummy_page_data = Mock()
        dummy_page_data.status_code = 200
        dummy_page_data.content = "unittest-page-content"
        stub_requests.get.return_value = dummy_page_data

        actual_value = stack_overflow._fetch_page(DUMMY_TARGET_URL)

        assert dummy_page_data.content == actual_value

    def test__fetch_page_should_raise_exception_for_error_in_fetching_page(self, mocker, stack_overflow: StackOverflow):
        stub_requests = mocker.patch(REQUESTS)
        stub_requests.get.side_effect = Exception("unittest-requests-get-exception")

        with pytest.raises(StackOverflowException):
            stack_overflow._fetch_page(DUMMY_TARGET_URL)

    def test__fetch_page_should_raise_exception_for_non_200_page_response(self, mocker, stack_overflow: StackOverflow):
        stub_requests = mocker.patch(REQUESTS)
        dummy_page_data = Mock()
        dummy_page_data.status_code = 500
        stub_requests.get.return_value = dummy_page_data

        with pytest.raises(StackOverflowException):
            stack_overflow._fetch_page(DUMMY_TARGET_URL)

    def test__make_soup_should_make_soup(self, stack_overflow: StackOverflow):
        expected_value = BeautifulSoup(stack_overflow_page_bytes_content, "html.parser")

        actual_value = stack_overflow._make_soup(stack_overflow_page_bytes_content)

        assert expected_value == actual_value

    def test__parse_page_should_parse_page_data(self, stack_overflow: StackOverflow, snapshot):
        soup = BeautifulSoup(stack_overflow_page_bytes_content, "html.parser")

        actual_value = stack_overflow._parse_page(soup)

        assert snapshot == actual_value

    def test__store_into_db_should_store_data_in_db(self, mocker, stack_overflow: StackOverflow):
        dummy_question_data = {"no": 1, "text": "unittest-text", "comments": "unittest-comments"}
        stack_overflow_question = StackOverflowQuestion(**dummy_question_data)
        dummy_answer_data = {"text": "unittest-text", "accepted": True}
        stack_overflow_answer = StackOverflowAnswer(**dummy_answer_data)
        questions_answers = [(stack_overflow_question, stack_overflow_answer)]
        spy_db = mocker.patch("src.external_sources.scrappers.stack_overflow.db")

        stack_overflow._store_into_db(questions_answers)

        spy_db.upsert_question_answer.assert_called_once_with(
            ques_no=stack_overflow_question.no,
            ques_text=stack_overflow_question.text,
            ques_comments=stack_overflow_question.comments,
            ans_text=stack_overflow_answer.text,
            ans_accepted=stack_overflow_answer.accepted,
        )

    def test_run(self, mocker, stack_overflow: StackOverflow):
        stub__fetch_page = mocker.patch("src.external_sources.scrappers.stack_overflow.StackOverflow._fetch_page")
        stub__make_soup = mocker.patch("src.external_sources.scrappers.stack_overflow.StackOverflow._make_soup")
        stub__parse_page = mocker.patch("src.external_sources.scrappers.stack_overflow.StackOverflow._parse_page")
        spy__store_into_db = mocker.patch("src.external_sources.scrappers.stack_overflow.StackOverflow._store_into_db")

        stack_overflow.run()

        page = stub__fetch_page(stack_overflow._target_url)
        soup = stub__make_soup(page)
        questions_answers = stub__parse_page(soup)
        spy__store_into_db.assert_called_once_with(questions_answers)
