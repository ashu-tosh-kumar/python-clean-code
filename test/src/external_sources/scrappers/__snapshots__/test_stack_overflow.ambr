# serializer version: 1
# name: TestStackOverflow.test__parse_page_should_parse_page_data
  list([
    tuple(
      StackOverflowQuestion(no=1, text='How to start with competitive coding from ZERO?', comments=None),
      StackOverflowAnswer(text='Read Medium Article: https://at-k.medium.com/how-to-start-with-competitive-coding-from-zero-cf99f3efdbb9', accepted=True),
    ),
    tuple(
      StackOverflowQuestion(no=2, text='How to develop a Wox Plugin using Python?', comments=None),
      StackOverflowAnswer(text='Read Medium Article: https://at-k.medium.com/how-to-develop-a-wox-plugin-using-python-8f2372281d7', accepted=True),
    ),
    tuple(
      StackOverflowQuestion(no=3, text='How to Streamline Your TODOs with TODO Notifier for Python projects?', comments=None),
      StackOverflowAnswer(text='Read Medium Article: https://at-k.medium.com/streamline-your-todos-with-todo-notifier-for-python-projects-6f95c03a2d34', accepted=True),
    ),
    tuple(
      StackOverflowQuestion(no=4, text='How to Write Impeccably Clean Code That Will Save Your Sanity?', comments=None),
      StackOverflowAnswer(text='Read Medium Article: <url>', accepted=True),
    ),
  ])
# ---
